core/src/main/java/org/apache/iceberg/TableUtil.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class TableUtil {
+  private TableUtil() {}
+
+  /**
+   * Returns the format version of the given table. Note that subclasses of {@link
+   * BaseMetadataTable} do not have a format version. The only exception to this is {@link
+   * PositionDeletesTable}, where the format version is fetched from the underlying table.
+   */
+  public static int formatVersion(Table table) {
+    Preconditions.checkArgument(null != table, "Invalid table: null");
+
+    // being able to read the format version from the PositionDeletesTable is mainly needed in
+    // SparkPositionDeletesRewrite when determining whether to rewrite V2 position deletes to DVs
+    if (table instanceof BaseMetadataTable) {
```

Comment:
I wish we had scala or even kotlin

Reply:
One of the reasons I want scala is I really want to just enumerate cases here. I would recommend we just go through all of our cases narrower then broader if we have exceptions so
```java
If (PositionDeleteTable)
  return format version 
else if (MetaTable) {
  Sorry Brah
}
else if (HasTableOperations) {
   return format version
}
else {
 // Sorry Brah
}
```
Although I am curious why position delete table needs a format version?

Reply:
> Although I am curious why position delete table needs a format version?

@RussellSpitzer this is mainly for `SparkPositionDeletesRewrite` (which operates against the position deletes table). Basically when rewriting existing position deletes we need to know whether we need to rewrite them to V2 position deletes or to DVs by looking at the underlying format version of the table. A table that was upgraded to V3 can still have V2 position deletes, meaning that these would then have to be rewritten as DVs

Reply:
Shouldn't we just be looking at the underlying table then? Shouldn't the converter look at the base table rather than the metadata table?

Ie

`formatVersion(metadataTable.baseTable)`

Rather than

`formatVersion(metadataTable)`

Implicitly calling `metadataTable.baseTable` but only sometimes

Reply:
> I would recommend we just go through all of our cases narrower then broader if we have exceptions so

I would prefer that too, but the fact that `SerializableMetadataTable` is a subclass of `SerializableTable` which in turn implements `HasTableOperations` makes this more difficult and you still need to differentiate there

Reply:
that would require changing a bunch of more places, because effectively we have a `Broadcast<Table>` in `SparkPositionDeletesRewrite`: https://github.com/apache/iceberg/blob/8351248150c1750dc17b7b0156eeea8427423ccd/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkPositionDeletesRewrite.java#L130-L131

Reply:
That is a an argument for ease of current implementation. But do you think it's the right decision going forward to have a format version for position deletes metadata table and none of the other metadata tables?

Reply:
I've simplified the checks a bit. I do think it makes sense to have an easy way to fetch the format version from the `PositionDeletesTable`, since it basically translates 1:1 to the underlying table and so the format version of the `PositionDeletesTable` would be the same as the underlying table . The question is whether we add this logic to `TableUtil` or to the calling site(s) and would like to hear what others think here /cc @amogh-jahagirdar @aokolnychyi